Enabling initial condition solving in dsolve

asmeurer opened the pull request:
**Example**

Single ODE:

```
In [7]: var("A")
Out[7]: A

In [8]: ics = {f(0): A}

In [9]: dsolve(f(x).diff(x) - f(x), f(x), ics=ics)
Out[9]:
          x
f(x) = A⋅ℯ
```

System:

```
In [10]: ics = {f(0): 1, g(0): 0}

In [11]: dsolve([f(x).diff(x) - f(x) + g(x), g(x).diff(x) - g(x) - f(x)], [f(x), g(x)], ics=ics)
Out[11]:
⎡        x                 x       ⎤
⎣f(x) = ℯ ⋅cos(x), g(x) = ℯ ⋅sin(x)⎦
```

Still TODO:
- [ ] Add tests
- [x] Add support for derivative initial conditions
- [ ] This uses `solve`. Is it possible to use `solveset`?
- [ ] Partial solutions? What do Maple and Mathematica do?

Fixes https://github.com/sympy/sympy/issues/4720

asmeurer committed:
Add initial version of solve_ics in ode.py

So far, it only works for single solutions, and non-derivative initial
conditions.

asmeurer committed:
Fix an error message about dsolve ics

asmeurer committed:
Extend solve_ics to work with multiple solutions (systems of ODEs)

asmeurer committed:
Change solve_ics to return the constant solution

The caller can then substitute these into the solution, or do something else
with them.

asmeurer committed:
Use a better variable name in solve_ics

asmeurer committed:
Enable initial condition solving from dsolve (both systems and single ODE)

asmeurer commented:
Even though this is still a work in progress, I would appreciate any comments from anyone who is familiar with the ODE module (which looks completely alien from what I originally wrote). CCing @smichr @cbm755

asmeurer commented:
CC @sympy/sympy-solvers regarding the `solve` vs. `solveset` thing.

asmeurer committed:
Don't try to solve for ics with power series hints

They do their own thing with initial conditions.

asmeurer commented:
@scopatz

scopatz commented on sympy/solvers/ode.py:
```diff
@@ -682,14 +686,65 @@ def _helper_simplify(eq, hint, match, simplify=True, **kwargs):
+            rv = rv1
+    return rv
+
+def solve_ics(sols, funcs, constants, ics):
```
Shouldn't this have a docstring?

scopatz replied:
done

scopatz commented:
Looks reasonable to me.  A couple of long lines; not sure how much I should care about that sort of thing.

asmeurer committed:
Update ics information in docstring of dsolve

asmeurer committed:
Add docstring to solve_ics()

asmeurer committed:
Add support for derivative initial conditions in solve_ics()

asmeurer committed:
Check for some additional cases of invalid ics in classify_ode()

asmeurer committed:
Add tests for invalid ics checking in classify_ode()

asmeurer committed:
Add tests that don't raise to test_classify_ode_ics()

asmeurer committed:
Add comment about NotImplementedError

asmeurer committed:
Grammar fix

cbm755 commented:
I'm not that familiar with it, although I started (and would still like to) fix systems (which are really terrible right now).

Anyway, I'll try to take a closer look....  My first impressions:

1.  is the dict approach to initial conditions ultimately limiting?  Maybe later we want general conditions like `Eq(sin(f(0)) = 1)`.  I think I'd rather see a list of `Eq` objects?

2.  its called "initial conditions" but at least two-point BVP seem to work with this approach.

3.  Thank you for doing this!

cbm755 commented:
My downstream tests pass except for one.  The failure is:
````
%!test
%! % Not enough initial conditions
%! syms y(x) C1
%! de = diff(y, 2) + 4*y == 0;
%! f = dsolve(de, y(0) == 3);
%! g = 3*cos(2*x) + C1*sin(2*x);
%! assert (isequal (rhs(f), g))    
````
I could probably prepare a SymPy patch with an `XFAIL` test for you for this...

asmeurer commented:
1. I'm not sure where those would come up, or how to even solve them. But at any rate, `{f(0): 1}` just means `f(0) = 1`, so you could always allow `{sin(f(0)): 1}`, if you ever wanted to add support for them. 

2. The API already exists, so I don't think it's worth changing. The docstring mentions that they are "initial/boundary conditions". I personally think it's unfortunate that two different names are used for what are effectively the same thing. 

3. Well I really wish that I had just done it back in 2009 when I wrote the rest of the module. It's not that complicated, especially if you are willing to accept non-answers in the cases where `solve` fails. 

Is there an issue summarizing the issues with systems. I've only tried basic ones, but they seem to work. 

> My downstream tests pass except for one. The failure is:

It wouldn't be hard to change this to return partial solutions. It's explicitly disallowed [here](https://github.com/sympy/sympy/pull/11264/files#diff-742912cfe279b1a8edf6f0ae25ba92b1R793), but only because I felt that it was better to fail in this case. What do Maple and Mathematica do in this case?

cbm755 commented:
@maprieto did our downstream implementation---at the time I encouraged upstreaming; but as you'll probably agree, SymPy's ODE stuff in not so easy to dig into!  My quick read of your code suggests its fairly similar in spirit.

nicoguaro commented:
Complementing the comment of @cbm755, I would mention another type of constraints that I find commonly, namely

1. Robin boundary conditions: ``Eq(a*diff(u(x), x).subs(x,x0) + b*u(x0), 0)``, that are common when you add a spring to a beam or when having natural convection.
2. Average values of the function or derivatives ``Eq(integrate(u(x), (x, a, b), c)``, these are common as weak boundary conditions in continuum mechanics. When the integral of the stress should equate a given force.
3. Normalization over the domain ``Eq(integrate(abs(u(x))**2, (x, -oo, oo), 1)``, common in quantum mechanics.
4. Continuity of the function or derivatives ``Eq(u1(x0), u2(x0))``, that appears when having two media in contact.

I think that those are the ones I have used more in the past. Is it too crazy to think about those already?

asmeurer commented:
I'm thinking that it might be better to have this ultimately implemented in `solve`, so that you can just do say `solve([Eq(f(x), C1*sin(x) + C2*cos(x)), Eq(f(0), 1), Eq(f(x).diff(x).subs(x, 0), 1)], [C1, C2])` and get the result `{C1: 1, C2: 1}`. And then for the other cases, you would also pass them in and solve would figure it out. 

But I don't plan on doing that here. I just want the most basic thing working, particularly initial conditions like f(x0) = a and f'(x0) = a.

I suppose it would be easy to allow `ics` could also be a list of equations instead of a dictionary as well (without breaking backwards compatibility with a dictionary, which is already part of the API).

nicoguaro commented:
I understand then.

I just tried to use it with the nonhomogeneous biharmonic equation: ``Eq(EI*diff(f(x), x, 4), q)``, defining the boundary conditions as described in the docstring.

    f = Function('f')
    x, EI, q, L = symbols('x EI q L')
    C1, C2, C3, C4 = symbols('C1 C2 C3 C4')
    eq = Eq(EI*diff(f(x), x, 4), q)
    sols = [dsolve(eq, f(x))]
    funcs = [f(x)]
    constants = [C1, C2, C3, C4]
    ics = {f(0): 0,
           f(x).diff(x).subs(x, 0): 0,
           f(x).diff(x, 2).subs(x, L): 0,
           f(x).diff(x, 3).subs(x, L): 0}
    solved_constants = solve_ics(sols, funcs, constants, ics)
    solved_constants

The result is

    ---------------------------------------------------------------------------
    ValueError                                Traceback (most recent call last)
    <ipython-input-59-d3cd734e8a33> in <module>()
    ----> 1 solved_constants = solve_ics(sols, funcs, constants, ics)

    <ipython-input-32-048254e999e4> in solve_ics(sols, funcs, constants, ics)
         51             S = diff_sols
         52         else:
    ---> 53             raise ValueError("Unrecognized initial condition")
         54 
         55         for sol in S:

    ValueError: Unrecognized initial condition

If I change it to 

    ics = {f(0): 0,
    Subs(diff(f(x), x), x, 0): 0,
    Subs(diff(f(x), x, 2), x, L): 0,
    Subs(diff(f(x), x, 3), x, L): 0}
    solved_constants = solve_ics(sols, funcs, constants, ics)
    solved_constants
    
It gives

    {C1: 0,
     C4: (EI*Derivative(f(L), L, L, L) - L*q)/(6*EI),
     C2: 0,
     C3: -L*Derivative(f(L), L, L, L)/2 + Derivative(f(L), L, L)/2 + L**2*q/(4*EI)}

And in none of the cases it works :-/

asmeurer commented:
Thanks for the report. I forgot that Derivative.subs can sometimes return a Derivative instead of a Subs. I'll need to fix the code to handle that.

asmeurer committed:
Update classify_ode to handle ics of type Derivative

Previously it only worked with Subs of Derivative, but f(x).diff(x).subs(x, y)
gives Derivative(f(y), y), not Subs(Derivative(f(x), x), x, 0), so it was
updated to handle this.

asmeurer committed:
Handle non-Subs Derivative ics in solve_ics()

asmeurer commented:
@nicoguaro pushed a commit that fixes that

```
In [23]: solved_constants
Out[23]:
⎧                   2             ⎫
⎪                  L ⋅q      -L⋅q ⎪
⎨C₁: 0, C₂: 0, C₃: ────, C₄: ─────⎬
⎪                  4⋅EI       6⋅EI⎪
⎩                                 ⎭
```

asmeurer committed:
Unrecognized initial conditions are NotImplementedError

asmeurer committed:
Add some tests for solve_ics()

asmeurer commented:
I've added some simple tests, but more complex tests are needed. I haven't checked how this handles situations where the constants appear in unusual places, (i.e., not just coefficients).

nicoguaro commented:
I could help with that, but I suck at branching/merging/pull requests.... Last time some of my changes got out of SymPy 1.0 because I mishandled the process.

asmeurer commented:
If it helps any, I won't be rebasing my branch.

asmeurer commented:
If anyone has access to Maple or Mathematica, it would help to check how they handle the [under specified case](https://github.com/sympy/sympy/pull/11264#issuecomment-228220956).

cbm755 commented:
Under-specified case in Maple:
````
> ode := diff(y(x),x,x) = 4*y(x);
                                   2
                                  d
                           ode := --- y(x) = 4 y(x)
                                    2
                                  dx

> dsolve(ode);                   
                      y(x) = _C1 exp(-2 x) + _C2 exp(2 x)

> dsolve({ode, y(0) = 1});       
                   y(x) = (1 - _C2) exp(-2 x) + _C2 exp(2 x)
````
I think this is natural: like (in fact it is) an underdetermined linear algebra problem.

asmeurer commented:
Maple even leaves the constant as C2 instead of C1 (then again, Maple never had constant simplification).

nicoguaro commented:
I tricked Maxima to have incomplete "boundary conditions", this is the result:

    (%i1) ode: 'diff(y,x,2) = 4*y;
                                        2
                                       d y
    (%o1)                              --- = 4 y
                                         2
                                       dx
    (%i2) sol: ode2(ode, y, x);
                                        2 x         - 2 x
    (%o6)                     y = %k1 %e    + %k2 %e

    (%i3) bc2(sol,x=0,y=0, x=0, y=0);
                                        - 2 x         2 x
    (%o7)                     y = %r1 %e      - %r1 %e

asmeurer committed:
Fix some issues with dsolve and initial conditions

- When simplify=False, some variables were not defined

- Constants for solve_ics should have been generated from the integrated
  solution. Otherwise, dummy variables would be included.

nicoguaro commented:
@asmeurer, I tried to get this branch of SymPy to help but failed XD...

In any case, I worked a couple of examples. One that works and one that fails, maybe they can help creating more tests.

### First example

We are solving ``diff(f(x),x) - f(x)*log(a*f(x))``

	a = symbols('a', positive=True)
	x0 = symbols('x0')
	eq = diff(f(x),x) - f(x)*sym.log(a*f(x))
	sol = dsolve(eq, f(x))
	solved_constants = solve_ics([sol], [f(x)], [C1], {f(0): x0})
	solved_constants

This gives 

	{C1: log(a) + log(x0)}

### Second example
This is the solution for ``diff(f(x), x, 2) + w**2*f(x) - w**2*cos(w*x)**3``.  I am writing the answer directly, since I am having problems with that equation. Check #11282.

	sol = Eq(f(x), C1*cos(C2 + w*x) + 3*w*x*sin(w*x)/8 + 3*cos(w*x)/16 - cos(3*w*x)/32)
	solved_constants = solve_ics([sol], [f(x)], [C1, C2], {f(0): 0, f(x).diff(x).subs(x,0): 0})

This gives

	---------------------------------------------------------------------------
	NotImplementedError                       Traceback (most recent call last)
	<ipython-input-50-0673979fc7f7> in <module>()
	----> 1 solved_constants = solve_ics([sol], [f(x)], [C1, C2], {f(0): 0, f(x).diff(x).subs(x,0): 0})

		730
		731     if len(solved_constants) > 1:
	--> 732         raise NotImplementedError("Initial conditions produced too many solutions for constants")
		733
		734     if len(solved_constants[0]) != len(constants):

	NotImplementedError: Initial conditions produced too many solutions for constants

But if we do it manually

	aux1 = sol.rhs.subs(x, 0)
	aux2 = sol.diff(x).rhs.subs(x, 0)

where

	[aux1, aux2]
	
is 

	[C1*cos(C2) + 5/32,  -C1*w*sin(C2)]

Then
	sym.solve([aux1, aux2], [C1, C2])

gives

	[(-5/32, 0), (5/32, pi)]

asmeurer commented:
Yeah, this is expected behavior. I didn't know how to handle the case when there is more than one solution set for the constants, so it raises NotImplementedError in that case. What output would you expect in the second case?

nicoguaro commented:
Well, for this example it is the same to pick any of the two solutions. They give the same solution at the end. The change of sign in one coefficient is collected in the change of sign of the other factor.

nicoguaro commented:
@asmeurer, since you mentioned that Robin-type of BCs/ICs are not implemented yet, I tried them expecting an error. Nevertheless, I obtained something else

    from __future__ import division
    from sympy import *
    x, m, Tb, T_inf, L, C1, C2, k = symbols('x, m T_b T_inf L C1 C2, k')
    Theta = symbols('Theta', cls=Function)
    ode = diff(Theta(x), x, 2) - m**2 * Theta(x)
    sol = dsolve(ode, Theta(x))
    ics =  {Theta(0): Tb - T_inf,
            diff(Theta(x), x).subs(x, L): -L/k*Theta(L)}
    constants = solve_ics([sol], [Theta(x)], [C1, C2], ics)
    constants

This is the output

      {C1: (L*Theta(L) + T_b*k*m*exp(L*m) - T_inf*k*m*exp(L*m))*exp(L*m)/(k*m*(exp(2*L*m) + 1)), C2: (-L*Theta(L)*exp(L*m) + k*m*(T_b - T_inf))/(k*m*(exp(2*L*m) + 1))}

If we try to solve for them

    eq1 = sol.subs([(Theta(x), Tb - T_inf), (x, 0)])
    eq2 = Eq(sol.rhs.subs(x, L), -k/L*sol.rhs.diff(x).subs(x, L))
    coef = solve([eq1, eq2], [C1, C2])

we obtain

    {C1: (L + k*m)*(T_b - T_inf)*exp(2*L*m)/(L*exp(2*L*m) - L + k*m*exp(2*L*m) + k*m), C2: -(L - k*m)*(T_b - T_inf)/(L*exp(2*L*m) - L + k*m*exp(2*L*m) + k*m)

I was expecting ``NotImplementedError: Unrecognized initial condition``.

asmeurer commented:
Good catch. That's technically right, but not very helpful, since the constants are given in terms of `Theta(L)`.

asmeurer committed:
Disallow ics where the point contains the function

These are not mathematically invalid, but the ics solver doesn't handle them
yet.

asmeurer commented:
@nicoguaro fixed. The type checking for initial conditions happens in `classify_ode`, so you'll only see the error if you use `dsolve(ics=ics)`. `solve_ics` will still give the wrong answer, because it assumes the ics have been checked by `classify_ode` already.

nicoguaro commented:
What about the case with multiple solutions for the system of equations?

asmeurer committed:
Handle degenerate cases in solve_ics

asmeurer committed:
Merge branch 'master' into dsolve-ics

asmeurer committed:
Preprocess symbols first in solve()

Otherwise, the inequality solver is called before symbol preprocessing, and
fails when symbols is a list of a single variable (like solve([x + 1 > 1],
[x])).

asmeurer committed:
Add test cases for ics when dsolve returns multiple solutions

asmeurer commented:
This is ready to be reviewed/merged.

asmeurer commented:
Test failure is https://github.com/sympy/sympy/issues/11684

asmeurer commented:
I went ahead and restarted the test.

asmeurer committed:
Merge branch 'master' into dsolve-ics

asmeurer committed:
Merge branch 'master' into dsolve-ics

asmeurer committed:
Merge branch 'master' into dsolve-ics

asmeurer commented:
We should get this merged. Just fixed the conflicts with master.

scopatz commented:
LGTM!
